Preview Flag not set with runtime only preview features

ref: https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3604#issuecomment-2609622313

Ref: JEP 12 https://openjdk.org/jeps/12 :
This issues is primarily about the point no: 2 (and maybe 3) from the following comment:
"If Java source code uses (i) a preview language feature of Java SE $N, or (ii) a class or interface declared using a preview language feature of Java SE $N, or (iii) a normal preview API of Java SE $N (whether referenced by name, or invoked, or overridden), then a Java compiler must record that the emitted class file depends on the preview features of Java SE $N."

From points 2 and 3 of https://openjdk.org/jeps/12: for features not affecting source, for eg: Scoped Values, there are two issues:
(1) ecj does not give an error if the preview feature is disabled while javac gives.
(2) Also, the minor version of 0xFFFF is not set in .class file even if the preview features is enabled.
eg:

```
public class X {
	public static void main(String[] args) {
		ScopedValue<Integer> si = ScopedValue.newInstance();
		System.out.println(si == null ? "hello" : "world");
	}
}

```

**Problem No:1**

use java24 (I used ea31) to compile without any flag 
$javac X.java
to get the following output:
"""
X.java:3: error: ScopedValue is a preview API and is disabled by default.
		ScopedValue<Integer> si = ScopedValue.newInstance();
		^
  (use --enable-preview to enable preview APIs)
X.java:3: error: ScopedValue is a preview API and is disabled by default.
		ScopedValue<Integer> si = ScopedValue.newInstance();
		                          ^
  (use --enable-preview to enable preview APIs)
2 errors
"""

Now use ecj (24) without --enable-preview flag, it just compiles (java used here is 24 ea31 just fyi)
$java -jar ./ecj.jar --release 24  X.java 
And we get X.class file
Expected Behaviour: ecj should throw the preview not enabled error

**Problem No:2**

use java24  to compile with --enable-preview flag 
$javac --enable-preview --release 24  X.java
X.class file is generated.
Inspect using javap,
$javap -p -v -c X|grep minor
we get ==>   _minor version: 65535_

Now, use ecj with the same set of flags:
$ java -jar ./ecj.jar --enable-preview --release 24  X.java
Inspect using javap,
we get ==>  _minor version: 0_

Expected Behaviour: ecj should have _minor version: 65535_

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/core/compiler/IProblem.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/core/compiler/IProblem.java
@@ -2250,6 +2250,8 @@ void setSourceStart(int sourceStart);
 	int JavaVersionNotSupported = Compliance + 1109;
 	/** @since 3.40*/
 	int JavaVersionTooRecent = Compliance + 1110;
+	/** @since 3.41 */
+	int PreviewAPIDisabled = Compliance + 1111;
 
 	/** @since 3.13 */
 	int UnlikelyCollectionMethodArgumentType = 1200;

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ASTNode.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ASTNode.java
@@ -58,6 +58,7 @@ import org.eclipse.jdt.internal.compiler.ASTVisitor;
 import org.eclipse.jdt.internal.compiler.ast.TypeReference.AnnotationPosition;
 import org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants;
 import org.eclipse.jdt.internal.compiler.env.AccessRestriction;
+import org.eclipse.jdt.internal.compiler.env.IBinaryAnnotation;
 import org.eclipse.jdt.internal.compiler.lookup.*;
 
 @SuppressWarnings({"rawtypes", "unchecked"})
@@ -460,12 +461,9 @@ public abstract class ASTNode implements TypeConstants, TypeIds {
 	public ASTNode concreteStatement() {
 		return this;
 	}
-	private void reportPreviewAPI(Scope scope, long modifiers) {
-		if (scope.compilerOptions().enablePreviewFeatures)
-			return;
-		if((modifiers & TagBits.AnnotationPreviewFeature) == TagBits.AnnotationPreviewFeature) {
-			scope.problemReporter().previewAPIUsed(this.sourceStart, this.sourceEnd,
-					(modifiers & TagBits.EssentialAPI) != 0);
+	private void reportPreviewAPI(Scope scope, IBinaryAnnotation previewAnnotation) {
+		if(previewAnnotation != null) {
+			scope.problemReporter().previewAPIUsed(scope, this.sourceStart, this.sourceEnd, previewAnnotation);
 		}
 	}
 	public final boolean isFieldUseDeprecated(FieldBinding field, Scope scope, int filteredBits) {
@@ -480,7 +478,7 @@ public abstract class ASTNode implements TypeConstants, TypeIds {
 			else
 				field.original().modifiers |= ExtraCompilerModifiers.AccLocallyUsed;
 		}
-		reportPreviewAPI(scope, field.tagBits);
+		reportPreviewAPI(scope, field.binaryPreviewAnnotation);
 
 		if ((field.modifiers & ExtraCompilerModifiers.AccRestrictedAccess) != 0) {
 			ModuleBinding module = field.declaringClass.module();
@@ -520,7 +518,7 @@ public abstract class ASTNode implements TypeConstants, TypeIds {
 	public final boolean isMethodUseDeprecated(MethodBinding method, Scope scope,
 			boolean isExplicitUse, InvocationSite invocation) {
 
-		reportPreviewAPI(scope, method.tagBits);
+		reportPreviewAPI(scope, method.binaryPreviewAnnotation);
 
 		// ignore references insing Javadoc comments
 		if ((this.bits & ASTNode.InsideJavadoc) == 0 && method.isOrEnclosedByPrivateType() && !scope.isDefinedInMethod(method)) {
@@ -601,7 +599,9 @@ public abstract class ASTNode implements TypeConstants, TypeIds {
 			// ignore cases where type is used from inside itself
 			((ReferenceBinding)refType.erasure()).modifiers |= ExtraCompilerModifiers.AccLocallyUsed;
 		}
-		reportPreviewAPI(scope, type.extendedTagBits);
+		if (type instanceof BinaryTypeBinding btb) {
+			reportPreviewAPI(scope, btb.binaryPreviewAnnotation);
+		}
 		if (refType.hasRestrictedAccess()) {
 			ModuleBinding module = refType.module();
 			LookupEnvironment env = (module == null) ? scope.environment() : module.environment;

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Annotation.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Annotation.java
@@ -376,16 +376,6 @@ public abstract class Annotation extends Expression {
 					}
 				}
 				break;
-			case TypeIds.T_JdkInternalPreviewFeature :
-				tagBits |= TagBits.AnnotationPreviewFeature;
-				for (MemberValuePair memberValuePair : memberValuePairs()) {
-					if (CharOperation.equals(memberValuePair.name, TypeConstants.ESSENTIAL_API)) {
-						if (memberValuePair.value instanceof TrueLiteral) {
-							tagBits |= TagBits.EssentialAPI;
-						}
-					}
-				}
-				break;
 			// marker annotations
 			case TypeIds.T_JavaLangDeprecated :
 				tagBits |= TagBits.AnnotationDeprecated;

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/ConstantPool.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/ConstantPool.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2024 IBM Corporation and others.
+ * Copyright (c) 2000, 2025 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -7,6 +7,10 @@
  * https://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
  *
  * Contributors:
  *     IBM Corporation - initial API and implementation
@@ -305,7 +309,10 @@ public class ConstantPool implements ClassFileConstants, TypeIds {
 	public static final char[] CloneSignature = "()Ljava/lang/Object;".toCharArray(); //$NON-NLS-1$
 	public static final char[] BOOTSTRAP = "bootstrap".toCharArray(); //$NON-NLS-1$
 	public static final char[] JAVA_LANG_RUNTIME_OBJECTMETHOD_BOOTSTRAP_SIGNATURE = "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/TypeDescriptor;Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/invoke/MethodHandle;)Ljava/lang/Object;".toCharArray(); //$NON-NLS-1$
-	public static final char[] PREVIEW_FEATURE = "/PreviewFeature".toCharArray(); //$NON-NLS-1$
+	public static final char[] PREVIEW_FEATURE = "Ljdk/internal/javac/PreviewFeature;".toCharArray(); //$NON-NLS-1$
+	public static final char[] REFLECTIVE = "reflective".toCharArray(); //$NON-NLS-1$ attribute of PreviewFeature
+	public static final char[] PREVIEW_FEATURE_JEP = "Ljdk/internal/javac/PreviewFeature$JEP;".toCharArray(); //$NON-NLS-1$
+	public static final char[] TITLE = "title".toCharArray(); //$NON-NLS-1$ attribute of the above annotation "JEP"
 	public static final char[] TYPESWITCH = "typeSwitch".toCharArray(); //$NON-NLS-1$
 	public static final char[] ENUMSWITCH = "enumSwitch".toCharArray(); //$NON-NLS-1$
 	public static final char[] JAVA_LANG_RUNTIME_SWITCHBOOTSTRAPS_SWITCH_SIGNATURE = "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;".toCharArray(); //$NON-NLS-1$

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BinaryTypeBinding.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BinaryTypeBinding.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2024 IBM Corporation and others.
+ * Copyright (c) 2000, 2025 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -7,6 +7,10 @@
  * https://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
  *
  * Contributors:
  *     IBM Corporation - initial API and implementation
@@ -105,6 +109,7 @@ public class BinaryTypeBinding extends ReferenceBinding {
 	protected LookupEnvironment environment;
 
 	protected SimpleLookupTable storedAnnotations = null; // keys are this ReferenceBinding & its fields and methods, value is an AnnotationHolder
+	public IBinaryAnnotation binaryPreviewAnnotation; // captures the exact preview feature of a preview API
 
 	private ReferenceBinding containerAnnotationType;
 	int defaultNullness = 0;
@@ -659,8 +664,8 @@ private void cachePartsFrom2(IBinaryType binaryType, boolean needFieldsAndMethod
 			}
 			for (IBinaryAnnotation annotation : declAnnotations) {
 				char[] typeName = annotation.getTypeName();
-				if (isPreviewFeature(typeName)) {
-					this.tagBits |= TagBits.AnnotationPreviewFeature;
+				if (CharOperation.equals(typeName, ConstantPool.PREVIEW_FEATURE)) {
+					this.binaryPreviewAnnotation = annotation;
 					break;
 				}
 			}
@@ -700,6 +705,7 @@ private void cachePartsFrom2(IBinaryType binaryType, boolean needFieldsAndMethod
 		this.environment.requestingType = previousRequester;
 	}
 }
+
 void markImplicitTerminalDeprecation(ReferenceBinding type) {
 	for (ReferenceBinding member : type.memberTypes()) {
 		member.tagBits |= TagBits.AnnotationTerminallyDeprecated;
@@ -858,16 +864,20 @@ private void createFields(IBinaryField[] iFields, IBinaryType binaryType, long s
 						? this.environment.getTypeFromSignature(binaryField.getTypeName(), 0, -1, false, this, missingTypeNames, walker)
 						: this.environment.getTypeFromTypeSignature(new SignatureWrapper(fieldSignature), Binding.NO_TYPE_VARIABLES, this, missingTypeNames, walker);
 					VariableBinding field = initialization.createBinding(this, binaryField, type);
+					boolean forceStoreAnnotations = false;
 					if (declAnnotations != null) {
 						for (IBinaryAnnotation annotation : declAnnotations) {
 							char[] typeName = annotation.getTypeName();
-							if (isPreviewFeature(typeName)) {
-								field.tagBits |= TagBits.AnnotationPreviewFeature;
+							if (CharOperation.equals(typeName, ConstantPool.PREVIEW_FEATURE) && field instanceof FieldBinding realField) {
+								realField.binaryPreviewAnnotation = annotation;
+								break;
+							} else if (CharOperation.equals(typeName, ConstantPool.PREVIEW_FEATURE_JEP)) {
+								forceStoreAnnotations = true;
 								break;
 							}
 						}
 					}
-					boolean forceStoreAnnotations = !this.environment.globalOptions.storeAnnotations
+					forceStoreAnnotations |= !this.environment.globalOptions.storeAnnotations
 							&& (this.environment.globalOptions.sourceLevel >= ClassFileConstants.JDK9
 							&& binaryField.getAnnotations() != null
 							&& (binaryField.getTagBits() & TagBits.AnnotationDeprecated) != 0);
@@ -904,14 +914,6 @@ private void createFields(IBinaryField[] iFields, IBinaryType binaryType, long s
 			initialization.setEmptyResult(this);
 		}
 	}
-}
-
-private boolean isPreviewFeature(char[] typeName) {
-	int index = CharOperation.lastIndexOf('/', typeName);
-	if (index != -1)
-		typeName = CharOperation.subarray(typeName, index, typeName.length);
-	index = CharOperation.indexOf(ConstantPool.PREVIEW_FEATURE, typeName, true);
-	return index == 0;
 }
 
 private MethodBinding createMethod(IBinaryMethod method, IBinaryType binaryType, long sourceLevel, char[][][] missingTypeNames) {
@@ -1099,8 +1101,8 @@ private MethodBinding createMethod(IBinaryMethod method, IBinaryType binaryType,
 	if (declAnnotations != null) {
 		for (IBinaryAnnotation annotation : declAnnotations) {
 			char[] typeName = annotation.getTypeName();
-			if (isPreviewFeature(typeName)) {
-				result.tagBits |= TagBits.AnnotationPreviewFeature;
+			if (CharOperation.equals(typeName, ConstantPool.PREVIEW_FEATURE)) {
+				result.binaryPreviewAnnotation = annotation;
 				break;
 			}
 		}

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/FieldBinding.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/FieldBinding.java
@@ -30,6 +30,7 @@ import org.eclipse.jdt.internal.compiler.ast.FakedTrackingVariable;
 import org.eclipse.jdt.internal.compiler.ast.FieldDeclaration;
 import org.eclipse.jdt.internal.compiler.ast.TypeDeclaration;
 import org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants;
+import org.eclipse.jdt.internal.compiler.env.IBinaryAnnotation;
 import org.eclipse.jdt.internal.compiler.impl.Constant;
 import org.eclipse.jdt.internal.compiler.problem.ProblemReporter;
 
@@ -39,6 +40,7 @@ public class FieldBinding extends VariableBinding {
 
 	public FakedTrackingVariable closeTracker;
 	public long extendedTagBits;
+	public IBinaryAnnotation binaryPreviewAnnotation; // captures the exact preview feature of a preview API
 
 protected FieldBinding() {
 	super(null, null, 0, null);

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/LookupEnvironment.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/LookupEnvironment.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2024 IBM Corporation and others.
+ * Copyright (c) 2000, 2025 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -7,6 +7,10 @@
  * https://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
  *
  * Contributors:
  *     IBM Corporation - initial API and implementation
@@ -1970,7 +1974,7 @@ ReferenceBinding getTypeFromConstantPoolName(char[] signature, int start, int en
 	return binding;
 }
 
-ReferenceBinding getTypeFromConstantPoolName(char[] signature, int start, int end, boolean isParameterized, char[][][] missingTypeNames) {
+public ReferenceBinding getTypeFromConstantPoolName(char[] signature, int start, int end, boolean isParameterized, char[][][] missingTypeNames) {
 	return getTypeFromConstantPoolName(signature, start, end, isParameterized, missingTypeNames, ITypeAnnotationWalker.EMPTY_ANNOTATION_WALKER);
 }
 

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/MethodBinding.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/MethodBinding.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2024 IBM Corporation and others.
+ * Copyright (c) 2000, 2025 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -7,6 +7,10 @@
  * https://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
  *
  * Contributors:
  *     IBM Corporation - initial API and implementation
@@ -49,6 +53,7 @@ import org.eclipse.jdt.internal.compiler.ast.TypeDeclaration;
 import org.eclipse.jdt.internal.compiler.ast.TypeReference.AnnotationPosition;
 import org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants;
 import org.eclipse.jdt.internal.compiler.codegen.ConstantPool;
+import org.eclipse.jdt.internal.compiler.env.IBinaryAnnotation;
 import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
 import org.eclipse.jdt.internal.compiler.util.Util;
 
@@ -65,6 +70,7 @@ public class MethodBinding extends Binding {
 	char[] signature;
 	public long tagBits;
 	public int extendedTagBits = 0; // See values in the interface ExtendedTagBits
+	public IBinaryAnnotation binaryPreviewAnnotation; // captures the exact preview feature of a preview API
 	// Used only for constructors
 	protected AnnotationBinding [] typeAnnotations = Binding.NO_ANNOTATIONS;
 

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ReferenceBinding.java
@@ -504,9 +504,6 @@ public void computeId() {
 				case 3: // only one type in this group, yet:
 					if (CharOperation.equals(TypeConstants.ORG_JUNIT_ASSERT, this.compoundName))
 						this.id = TypeIds.T_OrgJunitAssert;
-					else if (CharOperation.equals(TypeConstants.JDK_INTERNAL_PREVIEW_FEATURE, this.compoundName)
-							|| CharOperation.equals(TypeConstants.JDK_INTERNAL_JAVAC_PREVIEW_FEATURE, this.compoundName))
-						this.id = TypeIds.T_JdkInternalPreviewFeature;
 					return;
 				case 4:
 					if (!CharOperation.equals(TypeConstants.JAVA, packageName))

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TagBits.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TagBits.java
@@ -197,9 +197,6 @@ public interface TagBits {
 
 	long DefaultValueResolved = ASTNode.Bit60L;
 
-	long AnnotationPreviewFeature = ASTNode.Bit32L | ASTNode.Bit33L;
-	long EssentialAPI = ASTNode.Bit11;
-
 	// set when type contains non-private constructor(s)
 	long HasNonPrivateConstructor = ASTNode.Bit61L;
 

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TypeConstants.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TypeConstants.java
@@ -204,7 +204,6 @@ public interface TypeConstants {
 	char[][] JAVA_LANG_DEPRECATED = {JAVA, LANG, "Deprecated".toCharArray()}; //$NON-NLS-1$
 	char[] FOR_REMOVAL = "forRemoval".toCharArray(); //$NON-NLS-1$
 	char[] SINCE = "since".toCharArray(); //$NON-NLS-1$
-	char[] ESSENTIAL_API = "essentialAPI".toCharArray(); //$NON-NLS-1$
 	char[][] JAVA_LANG_ANNOTATION_DOCUMENTED = {JAVA, LANG, ANNOTATION, "Documented".toCharArray()}; //$NON-NLS-1$
 	char[][] JAVA_LANG_ANNOTATION_INHERITED = {JAVA, LANG, ANNOTATION, "Inherited".toCharArray()}; //$NON-NLS-1$
 	char[][] JAVA_LANG_ANNOTATION_REPEATABLE = {JAVA, LANG, ANNOTATION, "Repeatable".toCharArray()}; //$NON-NLS-1$

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TypeIds.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/TypeIds.java
@@ -151,8 +151,6 @@ public interface TypeIds {
 	// Java 14 preview
 	final int T_JavaLangRecord = 93;
 
-	final int T_JdkInternalPreviewFeature = 94;
-
 	// If you add new type id, make sure to bump up T_LastWellKnownTypeId if there is a cross over.
 	final int T_LastWellKnownTypeId = 128;
 

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/ProblemReporter.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/ProblemReporter.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2024 IBM Corporation and others.
+ * Copyright (c) 2000, 2025 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -100,9 +100,14 @@ import org.eclipse.jdt.internal.compiler.IErrorHandlingPolicy;
 import org.eclipse.jdt.internal.compiler.IProblemFactory;
 import org.eclipse.jdt.internal.compiler.ast.*;
 import org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants;
+import org.eclipse.jdt.internal.compiler.codegen.ConstantPool;
 import org.eclipse.jdt.internal.compiler.env.AccessRestriction;
+import org.eclipse.jdt.internal.compiler.env.EnumConstantSignature;
+import org.eclipse.jdt.internal.compiler.env.IBinaryAnnotation;
+import org.eclipse.jdt.internal.compiler.env.IBinaryElementValuePair;
 import org.eclipse.jdt.internal.compiler.env.ICompilationUnit;
 import org.eclipse.jdt.internal.compiler.flow.FlowInfo;
+import org.eclipse.jdt.internal.compiler.impl.BooleanConstant;
 import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
 import org.eclipse.jdt.internal.compiler.impl.Constant;
 import org.eclipse.jdt.internal.compiler.impl.IrritantSet;
@@ -9737,15 +9742,58 @@ public void previewFeatureUsed(int sourceStart, int sourceEnd) {
 			sourceStart,
 			sourceEnd);
 }
-public void previewAPIUsed(int sourceStart, int sourceEnd, boolean isFatal) {
-	this.referenceContext.compilationResult().usesPreview = true;
-	this.handle(
-			IProblem.PreviewAPIUsed,
-			NoArgument,
-			NoArgument,
-			isFatal ? ProblemSeverities.Error | ProblemSeverities.Fatal : ProblemSeverities.Warning,
-			sourceStart,
-			sourceEnd);
+public void previewAPIUsed(Scope scope, int sourceStart, int sourceEnd, IBinaryAnnotation previewAnnotation) {
+	String featureName = null; // FIXME: do we need a default string to use if the name is not found below?
+	boolean isReflective = false;
+	for (IBinaryElementValuePair valuePair : previewAnnotation.getElementValuePairs()) {
+		if (valuePair.getValue() instanceof EnumConstantSignature enumSig) {
+			// extract the feature title from the enum constant:
+			char[] typeName = enumSig.getTypeName();
+			ReferenceBinding enumType = scope.environment().getTypeFromConstantPoolName(typeName, 1, typeName.length-1, false, null);
+			if (enumType.isUnresolvedType())
+				enumType = (ReferenceBinding) BinaryTypeBinding.resolveType(enumType, scope.environment(), false);
+			FieldBinding field = enumType.getField(enumSig.getEnumConstantName(), true);
+			for (AnnotationBinding annotationBinding : field.getAnnotations()) {
+				if (CharOperation.equals(annotationBinding.getAnnotationType().constantPoolName(),
+						ConstantPool.PREVIEW_FEATURE_JEP, 1, ConstantPool.PREVIEW_FEATURE_JEP.length-1)) { // skip 'L' and ';'
+					for (ElementValuePair elementValuePair : annotationBinding.getElementValuePairs()) {
+						if (CharOperation.equals(ConstantPool.TITLE, elementValuePair.getName())
+								&& elementValuePair.value instanceof StringConstant constant) {
+							featureName = constant.stringValue();
+							break;
+						}
+					}
+				}
+			}
+		} else
+			if (CharOperation.equals(valuePair.getName(), ConstantPool.REFLECTIVE)) {
+			if (valuePair.getValue() instanceof BooleanConstant bool)
+				isReflective = bool.booleanValue();
+		}
+	}
+
+
+	String[] arguments = { featureName };
+	int problemId = -1;
+	int severity = -1;
+	if (!this.options.enablePreviewFeatures) {
+		problemId = IProblem.PreviewAPIDisabled;
+		severity = isReflective ? ProblemSeverities.Warning : ProblemSeverities.Error;
+	} else {
+		this.referenceContext.compilationResult().usesPreview = true;
+		if (this.options.isAnyEnabled(IrritantSet.PREVIEW)) {
+			severity = ProblemSeverities.Warning;
+			problemId = IProblem.PreviewAPIUsed;
+		}
+	}
+	if (problemId != -1 && severity != -1)
+		this.handle(
+				problemId,
+				arguments,
+				arguments,
+				severity,
+				sourceStart,
+				sourceEnd);
 }
 //Returns true if the problem is handled and reported (only errors considered and not warnings)
 private boolean validateRestrictedKeywords(char[] name, int start, int end, boolean reportSyntaxError) {

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/messages.properties
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/messages.properties
@@ -1,5 +1,5 @@
 ###############################################################################
-# Copyright (c) 2000, 2024 IBM Corporation and others.
+# Copyright (c) 2000, 2025 IBM Corporation and others.
 #
 # This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License 2.0
@@ -909,9 +909,10 @@
 1105 = The preview feature {0} is only available with source level {1} and above
 1106 = Preview features enabled at an invalid source release level {0}, preview can be enabled only at source level {1}
 1107 = The Java feature ''{0}'' is only available with source level {1} and above
-1108 = You are using an API that is part of a preview feature and may be removed in future
+1108 = You are using an API that is part of the preview feature ''{0}'' and may be removed in future
 1109 = Compiling for Java version ''{0}'' is no longer supported. Minimal supported version is ''{1}''
 1110 = Compiling for Java version ''{0}'' is not supported yet. Using ''{1}'' instead
+1111 = This API is part of the preview feature ''{0}'' which is disabled by default. Use --enable-preview to enable
 # more programming problems:
 1200 = Unlikely argument type {0} for {1} on a {2}
 1201 = Unlikely argument type for equals(): {0} seems to be unrelated to {2}

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/CompilerInvocationTests.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/CompilerInvocationTests.java
@@ -1253,6 +1253,7 @@ public void test011_problem_categories() {
 	    expectedProblemAttributes.put("PreviewFeaturesNotAllowed", new ProblemAttributes(CategorizedProblem.CAT_PREVIEW_RELATED));
 	    expectedProblemAttributes.put("FeatureNotSupported", new ProblemAttributes(CategorizedProblem.CAT_COMPLIANCE));
 	    expectedProblemAttributes.put("PreviewAPIUsed", new ProblemAttributes(CategorizedProblem.CAT_COMPLIANCE));
+	    expectedProblemAttributes.put("PreviewAPIDisabled", new ProblemAttributes(CategorizedProblem.CAT_COMPLIANCE));
 	    expectedProblemAttributes.put("JavaVersionNotSupported", new ProblemAttributes(CategorizedProblem.CAT_COMPLIANCE));
 	    expectedProblemAttributes.put("JavaVersionTooRecent", new ProblemAttributes(CategorizedProblem.CAT_COMPLIANCE));
 	    expectedProblemAttributes.put("SwitchExpressionsYieldIncompatibleResultExpressionTypes", new ProblemAttributes(CategorizedProblem.CAT_TYPE));
@@ -2395,6 +2396,7 @@ public void test012_compiler_problems_tuning() {
 	    expectedProblemAttributes.put("PreviewFeaturesNotAllowed", SKIP);
 	    expectedProblemAttributes.put("FeatureNotSupported", SKIP);
 	    expectedProblemAttributes.put("PreviewAPIUsed", SKIP);
+	    expectedProblemAttributes.put("PreviewAPIDisabled", SKIP);
 	    expectedProblemAttributes.put("JavaVersionNotSupported", SKIP);
 	    expectedProblemAttributes.put("JavaVersionTooRecent", SKIP);
 	    expectedProblemAttributes.put("SwitchExpressionsYieldIncompatibleResultExpressionTypes", SKIP);

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFeatureTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFeatureTest.java
@@ -125,10 +125,20 @@ public class PreviewFeatureTest extends AbstractRegressionTest9 {
 					"	^^^^\n" +
 					"Zork cannot be resolved to a type\n" +
 					"----------\n" +
-					"2. WARNING in X.java (at line 6)\n" +
+					"2. ERROR in X.java (at line 4)\n" +
+					"	ABC abc = null;\n" +
+					"	^^^\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
+					"----------\n" +
+					"3. ERROR in X.java (at line 6)\n" +
 					"	(new ABC()).doSomething();\n" +
 					"	^^^^^^^^^^^^^^^^^^^^^^^^^\n" +
-					"You are using an API that is part of a preview feature and may be removed in future\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
+					"----------\n" +
+					"4. ERROR in X.java (at line 6)\n" +
+					"	(new ABC()).doSomething();\n" +
+					"	     ^^^\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
 					"----------\n",
 					classLibs,
 					true,
@@ -167,6 +177,21 @@ public class PreviewFeatureTest extends AbstractRegressionTest9 {
 					"	Zork z = null;\n" +
 					"	^^^^\n" +
 					"Zork cannot be resolved to a type\n" +
+					"----------\n" +
+					"2. ERROR in X.java (at line 5)\n" +
+					"	ABC abc = null;\n" +
+					"	^^^\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
+					"----------\n" +
+					"3. ERROR in X.java (at line 7)\n" +
+					"	(new ABC()).doSomething();\n" +
+					"	^^^^^^^^^^^^^^^^^^^^^^^^^\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
+					"----------\n" +
+					"4. ERROR in X.java (at line 7)\n" +
+					"	(new ABC()).doSomething();\n" +
+					"	     ^^^\n" +
+					"This API is part of the preview feature \'Test Feature\' which is disabled by default. Use --enable-preview to enable\n" +
 					"----------\n",
 					classLibs,
 					true,
@@ -203,6 +228,21 @@ public class PreviewFeatureTest extends AbstractRegressionTest9 {
 					"	Zork z = null;\n" +
 					"	^^^^\n" +
 					"Zork cannot be resolved to a type\n" +
+					"----------\n" +
+					"2. WARNING in X.java (at line 4)\n" +
+					"	ABC abc = null;\n" +
+					"	^^^\n" +
+					"You are using an API that is part of the preview feature 'Test Feature' and may be removed in future\n" +
+					"----------\n" +
+					"3. WARNING in X.java (at line 6)\n" +
+					"	(new ABC()).doSomething();\n" +
+					"	^^^^^^^^^^^^^^^^^^^^^^^^^\n" +
+					"You are using an API that is part of the preview feature 'Test Feature' and may be removed in future\n" +
+					"----------\n" +
+					"4. WARNING in X.java (at line 6)\n" +
+					"	(new ABC()).doSomething();\n" +
+					"	     ^^^\n" +
+					"You are using an API that is part of the preview feature 'Test Feature' and may be removed in future\n" +
 					"----------\n",
 					classLibs,
 					true,

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFlagTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFlagTest.java
@@ -163,13 +163,67 @@ public class PreviewFlagTest extends AbstractRegressionTest9 {
 					"""
 			},
 			"----------\n" +
-			"1. ERROR in X.java (at line 12)\n" +
-			"ScopedValue<Integer> si = ScopedValue.newInstance();\n" +
-			"^\n" +
-			"ScopedValue is a preview feature and disabled by default. Use --enable-preview to enable\n" +
+			"1. ERROR in X.java (at line 3)\n" +
+			"	ScopedValue<Integer> si = ScopedValue.newInstance();\n" +
+			"	                          ^^^^^^^^^^^\n" +
+			"This API is part of the preview feature 'Scoped Values' which is disabled by default. Use --enable-preview to enable\n" +
 			"----------\n",
 			null,
 			true,
 			options);
+	}
+	public void testIssue3614_003() throws Exception {
+		Runner runner = new Runner();
+		runner.customOptions = getCompilerOptions(false);
+		runner.javacTestOptions = JavacTestOptions.DEFAULT; // don't enable preview
+		runner.testFiles = new String[] {
+				"X.java",
+				"""
+				import com.sun.source.tree.ImportTree;
+				public class X {
+					boolean foo(ImportTree tree) {
+						return tree.isModule();
+					}
+					public static void main(String... args) {}
+				}
+				"""
+			};
+		runner.expectedCompilerLog =
+			"----------\n" +
+			"1. WARNING in X.java (at line 4)\n" +
+			"	return tree.isModule();\n" +
+			"	       ^^^^^^^^^^^^^^^\n" +
+			"This API is part of the preview feature 'Module Import Declarations' which is disabled by default. Use --enable-preview to enable\n" +
+			"----------\n";
+		runner.runConformTest();
+	}
+	public void testIssue3614_003_enabled() throws Exception {
+		Runner runner = new Runner();
+		runner.customOptions = getCompilerOptions(true);
+		runner.vmArguments = VMARGS;
+		runner.javacTestOptions = JAVAC_OPTIONS;
+		runner.testFiles = new String[] {
+				"X.java",
+				"""
+				import com.sun.source.tree.ImportTree;
+				public class X {
+					boolean foo(ImportTree tree) {
+						return tree.isModule();
+					}
+					public void main(String... args) {
+						System.out.print(42);
+					}
+				}
+				"""
+			};
+		runner.expectedCompilerLog =
+			"----------\n" +
+			"1. WARNING in X.java (at line 4)\n" +
+			"	return tree.isModule();\n" +
+			"	       ^^^^^^^^^^^^^^^\n" +
+			"You are using an API that is part of the preview feature \'Module Import Declarations\' and may be removed in future\n" +
+			"----------\n";
+		runner.expectedOutputString = "42";
+		runner.runConformTest();
 	}
 }

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFlagTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PreviewFlagTest.java
@@ -1,0 +1,175 @@
+/*******************************************************************************
+ * Copyright (c) 2025 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * This is an implementation of an early-draft specification developed under the Java
+ * Community Process (JCP) and is made available for testing and evaluation purposes
+ * only. The code is not compatible with any specification of the JCP.
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.core.tests.compiler.regression;
+
+import java.io.IOException;
+import java.util.Map;
+import junit.framework.Test;
+import org.eclipse.jdt.core.util.ClassFileBytesDisassembler;
+import org.eclipse.jdt.core.util.ClassFormatException;
+import org.eclipse.jdt.internal.compiler.batch.FileSystem;
+import org.eclipse.jdt.internal.compiler.env.INameEnvironment;
+import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
+
+public class PreviewFlagTest extends AbstractRegressionTest9 {
+
+	private static final JavacTestOptions JAVAC_OPTIONS = new JavacTestOptions("--enable-preview -source 24");
+	private static final String[] VMARGS = new String[] {"--enable-preview"};
+	static {
+//		TESTS_NUMBERS = new int [] { 1 };
+//		TESTS_RANGE = new int[] { 1, -1 };
+//		TESTS_NAMES = new String[] { "testIssue3614_001" };
+//		TESTS_NAMES = new String[] { "testIssue3614_001" };
+	}
+	private String extraLibPath;
+	public static Class<?> testClass() {
+		return PreviewFlagTest.class;
+	}
+	public static Test suite() {
+		return buildMinimalComplianceTestSuite(testClass(), F_24);
+	}
+	public PreviewFlagTest(String testName) {
+		super(testName);
+	}
+
+	// ========= OPT-IN to run.javac mode: ===========
+	@Override
+	protected void setUp() throws Exception {
+		this.runJavacOptIn = true;
+		super.setUp();
+	}
+	@Override
+	protected void tearDown() throws Exception {
+		super.tearDown();
+		this.runJavacOptIn = false; // do it last, so super can still clean up
+	}
+	// =================================================
+
+	// Enables the tests to run individually
+	protected Map<String, String> getCompilerOptions(boolean preview) {
+		Map<String, String> defaultOptions = super.getCompilerOptions();
+		defaultOptions.put(CompilerOptions.OPTION_Compliance, CompilerOptions.VERSION_24);
+		defaultOptions.put(CompilerOptions.OPTION_Source, CompilerOptions.VERSION_24);
+		defaultOptions.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_24);
+		defaultOptions.put(CompilerOptions.OPTION_EnablePreviews, preview ? CompilerOptions.ENABLED : CompilerOptions.DISABLED);
+		defaultOptions.put(CompilerOptions.OPTION_ReportPreviewFeatures, CompilerOptions.WARNING);
+		return defaultOptions;
+	}
+
+	protected Map<String, String> getCompilerOptions() {
+		return getCompilerOptions(false);
+	}
+	protected String[] getDefaultClassPaths() {
+		String[] libs = DefaultJavaRuntimeEnvironment.getDefaultClassPaths();
+		if (this.extraLibPath != null) {
+			String[] l = new String[libs.length + 1];
+			System.arraycopy(libs, 0, l, 0, libs.length);
+			l[libs.length] = this.extraLibPath;
+			return l;
+		}
+		return libs;
+	}
+	@Override
+	protected INameEnvironment getNameEnvironment(final String[] testFiles, String[] classPaths, Map<String, String> options) {
+		this.classpaths = classPaths == null ? getDefaultClassPaths() : classPaths;
+		INameEnvironment[] classLibs = getClassLibs(false, options);
+		for (INameEnvironment nameEnvironment : classLibs) {
+			((FileSystem) nameEnvironment).scanForModules(createParser());
+		}
+		return new InMemoryNameEnvironment9(testFiles, this.moduleMap, classLibs);
+	}
+	@Override
+	protected void runConformTest(String[] testFiles, String expectedOutput) {
+		runConformTest(testFiles, expectedOutput, getCompilerOptions(true), VMARGS, JAVAC_OPTIONS);
+	}
+	@Override
+	protected void runConformTest(String[] testFiles, String expectedOutput, Map<String, String> customOptions) {
+		if(!isJRE22Plus)
+			return;
+		runConformTest(testFiles, expectedOutput, customOptions, VMARGS, JAVAC_OPTIONS);
+	}
+	protected void runNegativeTest(String[] testFiles, String expectedCompilerLog) {
+		Map<String, String> customOptions = getCompilerOptions(true);
+		Runner runner = new Runner();
+		runner.testFiles = testFiles;
+		runner.expectedCompilerLog = expectedCompilerLog;
+		runner.javacTestOptions = JAVAC_OPTIONS;
+		runner.customOptions = customOptions;
+		runner.expectedJavacOutputString = null;
+		runner.runNegativeTest();
+	}
+	class Runner extends AbstractRegressionTest.Runner {
+		public Runner(boolean reportPreview) {
+			this();
+			this.customOptions.put(CompilerOptions.OPTION_ReportPreviewFeatures, reportPreview ? CompilerOptions.WARNING : CompilerOptions.IGNORE);
+		}
+		public Runner() {
+			super();
+			this.vmArguments = VMARGS;
+			this.javacTestOptions = JAVAC_OPTIONS;
+			this.customOptions = getCompilerOptions();
+			this.customOptions.put(CompilerOptions.OPTION_EnablePreviews, CompilerOptions.ENABLED);
+		}
+	}
+	public void testIssue3614_001() throws IOException, ClassFormatException {
+		runConformTest(
+			new String[] {
+				"X.java",
+				"""
+					public class X {
+						public static void main(String[] args) {
+							ScopedValue<Integer> si = ScopedValue.newInstance();
+							System.out.println(si == null ? "hello" : "world");
+						}
+					}
+				"""
+			},
+			"world");
+		String expectedOutput =
+				"version 24 : 68.65535"
+				;
+		verifyClassFile(expectedOutput, "X.class", ClassFileBytesDisassembler.SYSTEM);
+	}
+	public void testIssue3614_002() throws IOException, ClassFormatException {
+		Map<String, String> options = getCompilerOptions();
+		options.put(CompilerOptions.OPTION_EnablePreviews, CompilerOptions.DISABLED);
+		runNegativeTest(new String[] {
+				"X.java",
+					"""
+					public class X {
+						public static void main(String[] args) {
+							ScopedValue<Integer> si = ScopedValue.newInstance();
+							System.out.println(si == null ? "hello" : "world");
+						}
+					}
+					"""
+			},
+			"----------\n" +
+			"1. ERROR in X.java (at line 12)\n" +
+			"ScopedValue<Integer> si = ScopedValue.newInstance();\n" +
+			"^\n" +
+			"ScopedValue is a preview feature and disabled by default. Use --enable-preview to enable\n" +
+			"----------\n",
+			null,
+			true,
+			options);
+	}
+}

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/TestAll.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/TestAll.java
@@ -263,6 +263,7 @@ public static Test suite() {
 	 since_24.add(ImplicitlyDeclaredClassesTest.class);
 	 since_24.add(PrimitiveInPatternsTest.class);
 	 since_24.add(PrimitiveInPatternsTestSH.class);
+	 since_24.add(PreviewFlagTest.class);
 	 // Add new tests for Java 24 here and/or move preview tests being moved from 23 to 24
 
 	 // Build final test suite

